### ISSUE TITLE
Cannot run the test suite under Python 3

### ISSUE THREAD

**Issue by jaraco** [jaraco/keyring#56]
I'm having trouble running the test suite under Python 3.2.  First, the zc.buildout bootstrap.py you suggest we download does not appear to be Python 3 compatible:

 python3.2 bootstrap 
  File "bootstrap", line 167
    exec ez_code in ez
               ^
SyntaxError: invalid syntax

I can run the test suite in Python 2.7 using `python -m unittest discover` but of course this doesn't work with Python 2.6.  I think it also doesn't work with Python 3 because the test suite itself needs to be run through 2to3.  You get stuff like this:

```
======================================================================
ERROR: keyring.tests.test_core (unittest.loader.ModuleImportFailure)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.2/unittest/case.py", line 370, in _executeTestPart
    function()
  File "/usr/lib/python3.2/unittest/loader.py", line 32, in testFailure
    raise exception
ImportError: Failed to import test module: keyring.tests.test_core
Traceback (most recent call last):
  File "/usr/lib/python3.2/unittest/loader.py", line 256, in _find_tests
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.2/unittest/loader.py", line 234, in _get_module_from_name
    __import__(name)
  File "/«PKGBUILDDIR»/keyring/__init__.py", line 9, in <module>
    from keyring.core import set_keyring, get_keyring, set_password, get_password
  File "/«PKGBUILDDIR»/keyring/core.py", line 15, in <module>
    from keyring import backend
  File "/«PKGBUILDDIR»/keyring/backend.py", line 114
    except OSError, e:
                  ^
SyntaxError: invalid syntax
```

So, exactly how do you run the test suite in Python 3.2?

---
- Bitbucket: https://bitbucket.org/kang/python-keyring-lib/issue/56
- Originally reported by: Barry Warsaw
- Originally created at: 2012-01-24T23:08:58.604

**Comment by jaraco** [comment]
Actually, the first recommended way to launch the tests is to use `python setup.py ptr`.
However, it still does fail with Python 3 since `keyring` now depends on [pyfs](https://code.google.com/p/pyfilesystem/) which latest release doesn't have Python 3 compatibility. Changes has been commited since then in the repository but no release have been done. I pinged maintainer at https://code.google.com/p/pyfilesystem/issues/detail?can=2&start=0&num=100&q=&colspec=ID%20Type%20Status%20Priority%20Milestone%20Owner%20Summary&groupby=&sort=&id=135

The error I'm getting now is the following:

```
$ python3 setup.py ptr        
zip_safe flag not set; analyzing archive contents...

Installed /tmp/easy_install-78myj3/pytest-runner-1.2/hgtools-2.0.2-py3.2.egg

Installed /home/jon/projects/externs/python-keyring-lib/pytest_runner-1.2-py3.2.egg
running ptr
Checking .pth file support in .
/usr/bin/python3 -E -c pass
Searching for mock
Reading http://pypi.python.org/simple/mock/
Reading http://www.voidspace.org.uk/python/mock/
Reading http://www.voidspace.org.uk/python/mock.html
Best match: mock 1.0.1
Downloading http://pypi.python.org/packages/source/m/mock/mock-1.0.1.zip#md5=869f08d003c289a97c1a6610faf5e913
Processing mock-1.0.1.zip
Running mock-1.0.1/setup.py -q bdist_egg --dist-dir /tmp/easy_install-rymhod/mock-1.0.1/egg-dist-tmp-xwreel
warning: no files found matching '*.png' under directory 'docs'
warning: no files found matching '*.css' under directory 'docs'
warning: no files found matching '*.html' under directory 'docs'
warning: no files found matching '*.js' under directory 'docs'
zip_safe flag not set; analyzing archive contents...

Installed /home/jon/projects/externs/python-keyring-lib/mock-1.0.1-py3.2.egg
Searching for fs
Reading http://pypi.python.org/simple/fs/
Reading http://code.google.com/p/pyfilesystem/
Reading http://code.google.com/p/pyfilesystem/downloads/list
Best match: fs 0.4.0
Downloading http://pyfilesystem.googlecode.com/files/fs-0.4.0.tar.gz
Processing fs-0.4.0.tar.gz
Running fs-0.4.0/setup.py -q bdist_egg --dist-dir /tmp/easy_install-09_epw/fs-0.4.0/egg-dist-tmp-iq5s2w
Traceback (most recent call last):
  File "setup.py", line 100, in <module>
    setup_mod.setup(**setup_params)
  File "/usr/lib/python3.2/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/usr/lib/python3.2/distutils/dist.py", line 917, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python3.2/distutils/dist.py", line 936, in run_command
    cmd_obj.run()
  File "/home/jon/projects/externs/python-keyring-lib/pytest_runner-1.2-py3.2.egg/ptr.py", line 50, in run
    self.distribution.fetch_build_eggs(self.distribution.tests_require)
  File "/usr/lib/python3/dist-packages/setuptools/dist.py", line 245, in fetch_build_eggs
    parse_requirements(requires), installer=self.fetch_build_egg
  File "/usr/lib/python3/dist-packages/pkg_resources.py", line 576, in resolve
    dist = best[req.key] = env.best_match(req, self, installer)
  File "/usr/lib/python3/dist-packages/pkg_resources.py", line 822, in best_match
    return self.obtain(req, installer) # try and download/install
  File "/usr/lib/python3/dist-packages/pkg_resources.py", line 834, in obtain
    return installer(requirement)
  File "/usr/lib/python3/dist-packages/setuptools/dist.py", line 294, in fetch_build_egg
    return cmd.easy_install(req)
  File "/usr/lib/python3/dist-packages/setuptools/command/easy_install.py", line 608, in easy_install
    return self.install_item(spec, dist.location, tmpdir, deps)
  File "/usr/lib/python3/dist-packages/setuptools/command/easy_install.py", line 638, in install_item
    dists = self.install_eggs(spec, download, tmpdir)
  File "/usr/lib/python3/dist-packages/setuptools/command/easy_install.py", line 828, in install_eggs
    return self.build_and_install(setup_script, setup_base)
  File "/usr/lib/python3/dist-packages/setuptools/command/easy_install.py", line 1105, in build_and_install
    self.run_setup(setup_script, setup_base, args)
  File "/usr/lib/python3/dist-packages/setuptools/command/easy_install.py", line 1094, in run_setup
    run_setup(setup_script, args)
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 31, in run_setup
    lambda: exec(compile(open(
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 73, in run
    return func()
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 34, in <lambda>
    {'__file__':setup_script, '__name__':'__main__'})
  File "setup.py", line 4, in <module>
    setup.py
  File "/tmp/easy_install-09_epw/fs-0.4.0/fs/__init__.py", line 25, in <module>
ImportError: No module named errors
```

---

Original comment by: Jonathan Ballet

**Comment by jaraco** [comment]
Waiting for https://code.google.com/p/pyfilesystem/issues/detail?can=2&start=0&num=100&q=&colspec=ID%20Type%20Status%20Priority%20Milestone%20Owner%20Summary&groupby=&sort=&id=135

---

Original comment by: Jonathan Ballet

**Comment by jaraco** [comment]
Explain how to launch the tests for Python 3

Fix #56

→ <<cset a816ca8d9856>>

---

Original comment by: Jonathan Ballet

**Comment by jaraco** [comment]
Well, I added a bit of documentation to explain how to launch the tests using Python 3, so it should fix the original problem reported by Barry. It's less than ideal, but it seems there's no better solution at the moment...

We still need to reactivate some of the backends and their tests when the corresponding modules will be Python 3-enabled.

---

Original comment by: Jonathan Ballet
